### ISSUE TITLE
Phase 0

### ISSUE THREAD

**Issue by ehartford** [cognitivecomputations/pytorch#1]
Phase 0: Essential Functions for Training & Inference
Tensor Creation & Manipulation

 torch_empty/zeros/ones/randn(shape, dtype, device, requires_grad)
 torch_from_blob(data, shape, dtype, requires_grad)
 torch_tensor_clone(tensor)
 torch_tensor_detach(tensor)
 torch_tensor_reshape(tensor, shape)
 torch_tensor_squeeze/unsqueeze(tensor, dim)

Basic Math Operations

 torch_add(a, b) / torch_sub/mul/div
 torch_matmul(a, b)
 torch_sum(tensor, dim, keepdim)
 torch_mean(tensor, dim, keepdim)
 torch_softmax(tensor, dim)
 torch_relu(tensor) / torch_sigmoid/tanh

Autograd (Critical for training)

 torch_tensor_set_requires_grad(tensor, bool)
 torch_tensor_requires_grad(tensor)
 torch_tensor_backward(tensor, gradient)
 torch_tensor_grad(tensor) - Access gradient
 torch_autograd_zero_grad() - Clear gradients
 torch_no_grad_guard_create/destroy() - Context manager

Optimizer Basics

 torch_optim_sgd_create(params, lr, momentum)
 torch_optim_adam_create(params, lr)
 torch_optim_step(optimizer)
 torch_optim_zero_grad(optimizer)
 torch_optim_free(optimizer)

Loss Functions

 torch_nn_mse_loss(input, target)
 torch_nn_cross_entropy_loss(input, target)
 torch_nn_bce_loss(input, target)

Model/Parameters

 torch_nn_linear_create(in_features, out_features)
 torch_nn_parameter_create(tensor)
 torch_module_parameters(module) - Get all parameters
 torch_module_forward(module, input)
 torch_save(tensor/module, filename)
 torch_load(filename)

Essential Properties

 torch_tensor_is_leaf(tensor)
 torch_tensor_is_contiguous(tensor)
 torch_tensor_contiguous(tensor)
 torch_tensor_to(tensor, device/dtype)
